### PR TITLE
Add configurations to enable or disable compression

### PR DESCRIPTION
**Pull request by riyafa**
## Automation tests
 - Integration tests
   > Added test case to verify enabled and disabled configuration.

### CODE REVIEW

File: tests/ballerina-test-integration/src/test/java/org/ballerinalang/test/service/http/sample/HttpServiceConfigTestCase.java [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+
+    private String compress(String str) throws Exception {
+        if (str == null || str.length() == 0) {
+            return str;
+        }
+        ByteArrayOutputStream obj = new ByteArrayOutputStream();
+        GZIPOutputStream gzip = new GZIPOutputStream(obj);
+        gzip.write(str.getBytes("UTF-8"));
+        gzip.close();
+        String outStr = obj.toString("UTF-8");
```
Comment (anupama-pathirage) [inline tests/ballerina-test-integration/src/test/java/org/ballerinalang/test/service/http/sample/HttpServiceConfigTestCase.java]: The variable outStr is redundant  here

File: tests/ballerina-test-integration/src/test/java/org/ballerinalang/test/service/http/sample/HttpServiceConfigTestCase.java [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+import org.testng.annotations.Test;
+
+import java.io.ByteArrayOutputStream;
+import java.io.File;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.zip.GZIPOutputStream;
+
+/**
+ * Test class for HTTP options request's content length and payload handling behavior.
```
Comment (chamil321) [inline tests/ballerina-test-integration/src/test/java/org/ballerinalang/test/service/http/sample/HttpServiceConfigTestCase.java]: This class description is not related to class name and it's functionality. Shall we change it?

File: tests/ballerina-test-integration/src/test/resources/httpService/httpServiceConfigTest.bal [hunk 3]
```diff
@@ -0,0 +1,34 @@
+import ballerina.net.http;
```
Comment (shafreenAnfar) [inline tests/ballerina-test-integration/src/test/resources/httpService/httpServiceConfigTest.bal]: Instead of having a integration test can we make it unit test ? So that we can avoid the transport. Transport already has a testcase for this.

File: stdlib/ballerina-http/src/main/java/org/ballerinalang/net/http/HttpUtil.java [hunk 4]
```diff
@@ -314,10 +315,27 @@ public static void prepareOutboundResponse(Context context, HTTPCarbonMessage in
...
 
+    /**
+     * Sets the content-encoding header to the outbound response based on the enabled http header.
+     * @param context The context
+     * @param outboundResponseMsg The inbound request message
+     */
+    private static void setCompressionHeader(Context context, HTTPCarbonMessage outboundResponseMsg) {
+        AnnAttachmentInfo configAnn = context.getServiceInfo().getAnnotationAttachmentInfo(
+                HttpConstants.PROTOCOL_PACKAGE_HTTP, HttpConstants.ANN_NAME_CONFIG);
+        AnnAttributeValue compressionEnabled = configAnn.getAttributeValue(
```
Comment (chamil321) [inline stdlib/ballerina-http/src/main/java/org/ballerinalang/net/http/HttpUtil.java]: configAnn can be null. Shall we add a null check?

File: stdlib/ballerina-http/src/main/java/org/ballerinalang/net/http/HttpUtil.java [hunk 5]
```diff
@@ -314,10 +315,27 @@ public static void prepareOutboundResponse(Context context, HTTPCarbonMessage in
 
         HttpUtil.checkEntityAvailability(context, outboundResponseStruct);
         HttpUtil.setKeepAliveHeader(context, outboundResponseMsg);
+        setCompressionHeader(context, outboundResponseMsg);
```
Comment (chamil321) [inline stdlib/ballerina-http/src/main/java/org/ballerinalang/net/http/HttpUtil.java]: Service annotation is accessed within setKeepAliveHeader method too. Better if you could move setKeepAliveHeader and setCompressionHeader to a common function.

File: tests/ballerina-test-integration/src/test/resources/httpService/httpServiceConfigTest.bal (hunk 3 above)
Comment (riyafa) [reply tests/ballerina-test-integration/src/test/resources/httpService/httpServiceConfigTest.bal]: Since this is a service configuration shouldn't we have an integration test? I think integration tests need to be written for other service configurations as well. Correct me if I am mistaken.

Comment (shafreenAnfar) [reply tests/ballerina-test-integration/src/test/resources/httpService/httpServiceConfigTest.bal]: Well, we don't have to. Because transport bit is tested at transport and the ballerina bit should be tested at ballerina. When testing the ballerina bit we shouldn't have to test the transport bit as well. Then there is a problem of separation of concerns :) .

What we are trying to archive is less number of integration tests and more number of unit tests similar to test-pyramid.

File: tests/ballerina-test/pom.xml [hunk 6]
```diff
@@ -84,6 +84,10 @@
             <groupId>org.testng</groupId>
             <artifactId>testng</artifactId>
         </dependency>
+        <dependency>
+            <groupId>org.mockito</groupId>
+            <artifactId>mockito-all</artifactId>
```
Comment (shafreenAnfar) [inline tests/ballerina-test/pom.xml]: Don't we have to say it is a test phase dependency ?

File: tests/ballerina-test/src/test/java/org/ballerinalang/test/services/configuration/CompressionConfigurationTest.java [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+        AnnAttributeValue annAttributeValue = mock(AnnAttributeValue.class);
+
+        when(context.getServiceInfo()).thenReturn(serviceInfo);
+        when(serviceInfo
+                     .getAnnotationAttachmentInfo(HttpConstants.PROTOCOL_PACKAGE_HTTP, HttpConstants.ANN_NAME_CONFIG))
+                .thenReturn(annAttachmentInfo);
+        when(annAttachmentInfo.getAttributeValue(HttpConstants.ANN_CONFIG_ATTR_COMPRESSION_ENABLED)).thenReturn(
+                annAttributeValue);
+        when(annAttributeValue.getBooleanValue()).thenReturn(false);
+
```
Comment (shafreenAnfar) [inline tests/ballerina-test/src/test/java/org/ballerinalang/test/services/configuration/CompressionConfigurationTest.java]: There are some additional spacing ..

File: tests/ballerina-test/src/test/java/org/ballerinalang/test/services/configuration/CompressionConfigurationTest.java [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+        AnnAttributeValue annAttributeValue = mock(AnnAttributeValue.class);
+
+        when(context.getServiceInfo()).thenReturn(serviceInfo);
+        when(serviceInfo
+                     .getAnnotationAttachmentInfo(HttpConstants.PROTOCOL_PACKAGE_HTTP, HttpConstants.ANN_NAME_CONFIG))
+                .thenReturn(annAttachmentInfo);
+        when(annAttachmentInfo.getAttributeValue(HttpConstants.ANN_CONFIG_ATTR_COMPRESSION_ENABLED)).thenReturn(
+                annAttributeValue);
+        when(annAttributeValue.getBooleanValue()).thenReturn(true);
+
```
Comment (shafreenAnfar) [inline tests/ballerina-test/src/test/java/org/ballerinalang/test/services/configuration/CompressionConfigurationTest.java]: There are some additional spacing ...

File: tests/ballerina-test/pom.xml (hunk 6 above)
Comment (riyafa) [reply tests/ballerina-test/pom.xml]: added the test scope

File: tests/ballerina-test/src/test/java/org/ballerinalang/test/services/configuration/CompressionConfigurationTest.java (hunk 8 above)
Comment (riyafa) [reply tests/ballerina-test/src/test/java/org/ballerinalang/test/services/configuration/CompressionConfigurationTest.java]: Removed spacing

File: tests/ballerina-test/src/test/java/org/ballerinalang/test/services/configuration/CompressionConfigurationTest.java (hunk 7 above)
Comment (riyafa) [reply tests/ballerina-test/src/test/java/org/ballerinalang/test/services/configuration/CompressionConfigurationTest.java]: Removed spacing
